### PR TITLE
Fix S1186 FP: Empty method comes from interface

### LINKED ISSUE #7629: Fix S1186 FP: Empty method comes from interface
### Description

S1186 is reported for empty methods that implement interface

### Repro steps

```c#
interface IFilter
{
    void OnEnter();

    void OnExit();
}

class Filter : IFilter
{
    public void OnEnter() // <- is reported here
    {
    }

    public void OnExit()
    {
        Console.WriteLine("exiting");
    }
}
```

### Expected behavior

No diagnostic is reported

### Actual behavior

Diagnostic is reported

### Related information

* C#/VB.NET Plugins version 9.5.0.73987
* Visual Studio version latest preview
* MSBuild / dotnet version .NET 8 preview 6

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #7629

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Extensions/ISymbolExtensions.cs [hunk 1]
```diff
@@ -48,5 +48,22 @@ internal static class ISymbolExtensions
         // https://www.jetbrains.com/help/resharper/Reference__Code_Annotation_Attributes.html#NotNullAttribute
         private static bool IsNotNullAttribute(AttributeData attribute) =>
             attribute.HasAnyName("ValidatedNotNullAttribute", "NotNullAttribute");
+
+        // https://github.com/dotnet/roslyn/blob/2a594fa2157a734a988f7b5dbac99484781599bd/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs#L93
+        public static ImmutableArray<ISymbol> ExplicitOrImplicitInterfaceImplementations(this ISymbol symbol)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/ISymbolExtensions.cs:53]: I have created this trello card in How we work about the placement of the helpers:
https://trello.com/c/nw3yLmEY
See also my comment about coverage there.

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Extensions/ISymbolExtensions.cs:53]: Consider also adding the complementing methods that are available in Roslyn, like `ImplicitInterfaceImplementations` and `ExplicitInterfaceImplementations`. We should only do this after the trello card is resolved.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EmptyMethod.cs [hunk 2]
```diff
@@ -57,7 +57,8 @@ protected override void CheckMethod(SonarSyntaxNodeReportingContext context)
 
         private static bool ShouldMethodBeExcluded(SonarSyntaxNodeReportingContext context, BaseMethodDeclarationSyntax methodNode) =>
             methodNode.Modifiers.Any(SyntaxKind.VirtualKeyword)
-            || context.SemanticModel.GetDeclaredSymbol(methodNode) is { IsOverride: true, OverriddenMethod.IsAbstract: true }
+            || (context.SemanticModel.GetDeclaredSymbol(methodNode) is var symbol
+                && (symbol is { IsOverride: true, OverriddenMethod.IsAbstract: true } || symbol.ExplicitOrImplicitInterfaceImplementations().Any()))
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EmptyMethod.cs:61]:
```suggestion
                && (symbol is { IsOverride: true, OverriddenMethod.IsAbstract: true } || !symbol.ExplicitOrImplicitInterfaceImplementations().IsEmpty))
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Some small changes and a question about the location of the new extension method.

**Review by martin-strecker-sonarsource — COMMENTED** [review]
LGTM
The changes in ITs seem reasonable.

**Comment by cristian-ambrosini-sonarsource** [comment]
Peach validation:
- 2603 lost FPs. I checked randomly more than 50 and they were all methods coming from an interface.

**Pull request merged**